beetsplug/lastgenre/__init__.py

```diff
@@ -111,39 +111,42 @@ def setup(self):
             self.import_stages = [self.imported]
 
         self._genre_cache = {}
+        self.whitelist = self._load_whitelist()
+        self.c14n_branches, self.canonicalize = self._load_c14n_tree()
 
-        # Read the whitelist file if enabled.
-        self.whitelist = set()
+    def _load_whitelist(self) -> set[str]:
+        whitelist = set()
         wl_filename = self.config["whitelist"].get()
         if wl_filename in (True, ""):  # Indicates the default whitelist.
             wl_filename = WHITELIST
         if wl_filename:
             wl_filename = normpath(wl_filename)
             with open(wl_filename, "rb") as f:
-                for line in f:
-                    line = line.decode("utf-8").strip().lower()
+                for raw_line in f:
+                    line = raw_line.decode("utf-8").strip().lower()
                     if line and not line.startswith("#"):
-                        self.whitelist.add(line)
+                        whitelist.add(line)
+        return whitelist
 
-        # Read the genres tree for canonicalization if enabled.
-        self.c14n_branches = []
+    def _load_c14n_tree(self) -> tuple[list[list[str]], bool]:
+        c14n_branches: list[list[str]] = []
         c14n_filename = self.config["canonical"].get()
-        self.canonicalize = c14n_filename is not False
-
+        canonicalize = c14n_filename is not False
         # Default tree
         if c14n_filename in (True, ""):
             c14n_filename = C14N_TREE
-        elif not self.canonicalize and self.config["prefer_specific"].get():
+        elif not canonicalize and self.config["prefer_specific"].get():
             # prefer_specific requires a tree, load default tree
             c14n_filename = C14N_TREE
-
         # Read the tree
         if c14n_filename:
             self._log.debug("Loading canonicalization tree {0}", c14n_filename)
-            c14n_filename = normpath(c14n_filename)
-            with codecs.open(c14n_filename, "r", encoding="utf-8") as f:
+            with codecs.open(
+                str(normpath(c14n_filename)), "r", encoding="utf-8"
+            ) as f:
```

Comment:
Had a mypy error here that only happened in the CI, not locally:
```
beetsplug/lastgenre/__init__.py:144: error: Argument 1 to "open" has incompatible type "bytes"; expected "str"  [arg-type]
```
 Seems like codecs.open wants str., normpath returns bytes though. normpath is essential here since it does all kinds of sanity checks and transforms so I'd like to keep it. Wrapping with str() is the most straightforward way I suppose? 
 
 Another way that works would be to wrap with syspath(), would this be better / more "beetonic"? @wisp3rwind you are always a good address with such questions ;-) thanks in advance! :-)

Reply:
I think I answered this myself. Yes syspath(normpath()) is the way to go here. I had these errors in the CI:

```
FAILED test/plugins/test_lastgenre.py::LastGenrePluginTest::test_c14n_only - FileNotFoundError: [Errno 2] No such file or directory: "b'/home/runner/work/beets/beets/beetsplug/lastgenre/genres-tree.yaml'"
```

which are now fixed. Sorry the bother and nevermind! Resolved.